suddi/claudia-local-api
# Feature/binary response support

## Overview
[Claudia.JS can handle serving up binary content through `claudia-api-builder`](https://claudiajs.com/tutorials/binary-content.html), [as can express.](https://expressjs.com/en/api.html#res.send) However, if you follow Claudia's API builder binary content handling tutorial to serve a `.png` using `claudia-local-api`, you are met with something that looks like:
![image](https://user-images.githubusercontent.com/16990679/42578035-b0c33c9c-84f3-11e8-992e-ae9182be238e.png) (Firefox) or 
![image](https://user-images.githubusercontent.com/16990679/42578066-c1f284c8-84f3-11e8-9d01-7bf89aee6c9c.png) (Chrome).

The root of the issue seems to be that whenever you add the binary content handling configuration object
```
{ 
  requestContentHandling: 'CONVERT_TO_TEXT', 
  success: { 
    contentType: 'image/png', 
    contentHandling: 'CONVERT_TO_BINARY' 
  } 
}
```
 to the `claudia-api-builder` handler, the api builder properly adds the `"Content-Type": "image/png"` header to the response but does not send the content as actual binary data. This can be seen in the response object that `claudia-local-api` logs (I've logged `typeof response.body` for reference and snipped out most of the body):
```
14:18:12.381Z  INFO claudia-local-api:
    {
        "statusCode": 201,                                                                                                                                                    
        "headers": {                                                                                                                                                          
            "Content-Type": "image/png",                                                                                                                                      
            "Access-Control-Allow-Origin": "*",                                                                                                                               
            "Access-Control-Allow-Headers": "Content-Type,Authorization,X-Amz-Date,X-Api-Key,X-Amz-Security-Token",                                                           
            "Access-Control-Allow-Methods": "GET,OPTIONS",                                                                                                                    
            "Access-Control-Allow-Credentials": "true",                                                                                                                       
            "Access-Control-Max-Age": "0",                                                                                                                                    
            "called": "handleBinaryRequest"                                                                                                                                   
        },                                                                                                                                                                    
        "body": "iVBORw0KGgoAAAANSUhEUgAAAfQAAAH0CAYAAADL1t+KAAAAAXNSR0IArs4c6QAAAAZiS0dEAP8[...]duYAAAAldEVYdGRhdGU6bW9kaWZ5ADIwMTgtMDYtMjdUMTI6NDU6MzcrMDE6MDBeSRAhAAAAAElFTkSuQmCC",
        "isBase64Encoded": true
    }
14:18:12.383Z  INFO claudia-local-api: "string"
```
The response body _is_ a utf-8-encoded image, but it's sent as a string instead of as a Buffer, which causes a conflict with the response headers that leads to this rendering error.

This PR adds in a function `parseContentHandling` which roots around in the api config object/route params in order to fetch the current value of `success.contentHandling`. If this is set by the user to `CONVERT_TO_BINARY`, `response.body` is wrapped in a base-64 Buffer in the express response object.

### Demo
![image](https://user-images.githubusercontent.com/16990679/42579554-98aa3fc2-84f6-11e8-801d-e01dc986e9dd.png)
![image](https://user-images.githubusercontent.com/16990679/42579385-39cdd0ea-84f6-11e8-8fbf-2c2771e41440.png)


## Testing Instructions

 * The original error is reproducible by following Claudia.js's tutorial [here](https://claudiajs.com/tutorials/binary-content.html). You need to set up a `claudia-api-builder` endpoint that serves up some sort of binary content (a .png image in the example I included in this PR), and has a handling configuration object that requests converting to binary with a content type of "image/png".
* Running the same test app with the changes from this PR should allow your binary content to display properly.
